kvserver: estimate AddSStable stats if ComputeStatsDiff assuptions violated

msbutler opened the pull request:
Previously the request would fail-- now we internally retry the stats computation with estimates. We should further investigate why the ComputeStatsDiff assumptions are violated in the c2c/kv0 test though.

Fixes #151979
Release note: none

msbutler committed:
kvserver: estimate AddSStable stats if ComputeStatsDiff assuptions violated

Previously the request would fail-- now we internally retry the stats
computation with estimates. We should further investigate why the
ComputeStatsDiff assumptions are violated in the c2c/kv0 test though.

Fixes #151979
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152450)
<!-- Reviewable:end -->

jeffswenson commented on pkg/kv/kvserver/batcheval/cmd_add_sstable.go:
```diff
@@ -474,12 +474,11 @@ func computeSSTStatsDiffWithFallback(
-		// Fall back to stats estimates if there are range keys in the engine.
-		log.VEventf(ctx, 2, "computing SST stats as estimates after detecting range keys in engine")
+	if errors.IsAny(err, storage.ComputeSSTStatsDiffReaderHasRangeKeys, storage.ComputeStatsDiffViolation) {
+		log.Infof(ctx, "computing SST stats as estimates because of ComputeSSTStatsDiff error: %s", err)
```
This is probably rare enough we could make it a warning.

jeffswenson reviewed (approved):
LGTM

msbutler committed:
kv/bulk: turn off ComputeStatsDiff by default

We should figure out how often assumptions are violated in c2c/kv0 before
turning this back on by default.

Epic: none

Release note: none
